CSLabs.Api/Services/TestVmConnectionService.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System.Linq;
+using System.Threading.Tasks;
+using CSLabs.Api.Models;
+using CSLabs.Api.Proxmox;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+
+namespace CSLabs.Api.Services
+{
+    public class TestVmConnectionService
+    {
+        private DefaultContext Context { get; }
+
+        private ProxmoxManager ProxmoxManager;
+        
+        public TestVmConnectionService(DefaultContext context)
+        {
+            Context = context;
+        }
+        
+        public async Task<IActionResult> TestLabVmConnection()
+        {
+            var hypervisors = await Context.Hypervisors
+                .Include(h => h.HypervisorNodes)
+                .ToListAsync();
+        
+            var labVms = await Context.LabVms
+                .Include(v => v.Id)
+                .ToListAsync();
+
+            foreach (var hypervisor in hypervisors) 
+            {
+                var api = ProxmoxManager.GetProxmoxApi(hypervisor.HypervisorNodes.First());
+                
+                foreach (var labVm in labVms) 
+                { 
+                    try 
+                    { 
+                        var vmStatus = await api.GetVmStatus(labVm.Id); 
+                        if (vmStatus.IsStopped())  // vm is down
+                        { 
+                            // attempt to restart the vm
+                            try
+                            {
+                                await api.StartVM(labVm.Id); // 1st attempt
+                            }
+                            catch (ProxmoxRequestException)
+                            {
+                                try
+                                {
+                                    await api.StartVM(labVm.Id); // 2nd attempt
+                                }
+                                catch (ProxmoxRequestException)
+                                {
+                                    try
+                                    {
+                                        //third attempt
+                                        await api.ShutdownVm(labVm.Id);
```

Comment:
I would refactor this to avoid heavy nesting. If you are going to try and make multiple attemps refactor it into a method with the attempt number passed as a parameter with a single try catch that calls itself recursively